include/pybind11/detail/internals.h

```diff
@@ -401,9 +401,30 @@ inline void translate_local_exception(std::exception_ptr p) {
 }
 #endif
 
+inline object get_internals_state_dict() {
+    object state_dict;
+#if PY_VERSION_HEX < 0x03080000 || defined(PYPY_VERSION)
```

Comment:
It turns out it is extremely easy to invite the Python 3.12 core developers to test with pybind11, with a 1 or 2 line change here, depending on how you count. Diff below.

I tested that diff with a local installation of Python 3.12.0a1. Log also below.


```diff
diff --git a/include/pybind11/detail/internals.h b/include/pybind11/detail/internals.h
index 03c10506..c8ac22e7 100644
--- a/include/pybind11/detail/internals.h
+++ b/include/pybind11/detail/internals.h
@@ -403,7 +403,8 @@ inline void translate_local_exception(std::exception_ptr p) {

 inline object get_internals_state_dict() {
     object state_dict;
-#if PY_VERSION_HEX < 0x03080000 || defined(PYPY_VERSION)
+#if (PYBIND11_INTERNALS_VERSION <= 4 && PY_VERSION_HEX < 0x030C0000)                              \
+    || PY_VERSION_HEX < 0x03080000 || defined(PYPY_VERSION)
     state_dict = reinterpret_borrow<object>(PyEval_GetBuiltins());
 #else
 #    if PY_VERSION_HEX < 0x03090000
```

```
+ /usr/local/google/home/rwgk/usr_local_like/Python-3.12.0a1/bin/python3 /usr/local/google/home/rwgk/clone/pybind11_scons/run_tests.py ../pybind11
Running tests in directory "/usr/local/google/home/rwgk/forked/pybind11/tests/test_embed":
===============================================================================
All tests passed (1554 assertions in 12 test cases)

Running tests in directory "/usr/local/google/home/rwgk/forked/pybind11/tests":
=========================================================== test session starts ============================================================
platform linux -- Python 3.12.0a1, pytest-7.2.0, pluggy-1.0.0
C++ Info: Debian Clang 14.0.6 C++17 __pybind11_internals_v4_clang_libstdcpp_cxxabi1002__
rootdir: /usr/local/google/home/rwgk/forked/pybind11/tests, configfile: pytest.ini
collected 698 items

test_async.py ..                                                                                                                     [  0%]
test_buffers.py .........                                                                                                            [  1%]
test_builtin_casters.py ....................                                                                                         [  4%]
test_call_policies.py ........                                                                                                       [  5%]
test_callbacks.py ............                                                                                                       [  7%]
test_chrono.py ...........................................                                                                           [ 13%]
test_class.py ...............................                                                                                        [ 17%]
test_const_name.py ......................                                                                                            [ 21%]
test_constants_and_functions.py .....                                                                                                [ 21%]
test_copy_move.py ........                                                                                                           [ 22%]
test_custom_type_casters.py ...                                                                                                      [ 23%]
test_custom_type_setup.py ..                                                                                                         [ 23%]
test_docstring_options.py .                                                                                                          [ 23%]
test_eigen_matrix.py .....................ss.......                                                                                  [ 28%]
test_eigen_tensor.py .............................................................................................................   [ 43%]
test_enum.py .........                                                                                                               [ 44%]
test_eval.py ....                                                                                                                    [ 45%]
test_exceptions.py ......................                                                                                            [ 48%]
test_factory_constructors.py ...............                                                                                         [ 50%]
test_gil_scoped.py .....                                                                                                             [ 51%]
test_iostream.py ......................                                                                                              [ 54%]
test_kwargs_and_defaults.py ........                                                                                                 [ 55%]
test_local_bindings.py ..........                                                                                                    [ 57%]
test_methods_and_attributes.py ......................                                                                                [ 60%]
test_modules.py .......                                                                                                              [ 61%]
test_multiple_inheritance.py ..................                                                                                      [ 64%]
test_numpy_array.py .........................................................                                                        [ 72%]
test_numpy_dtypes.py ...............                                                                                                 [ 74%]
test_numpy_vectorize.py ........                                                                                                     [ 75%]
test_opaque_types.py ...                                                                                                             [ 75%]
test_operator_overloading.py .....                                                                                                   [ 76%]
test_pickling.py ........                                                                                                            [ 77%]
test_pytypes.py ..................................................................................                                   [ 89%]
test_sequences_and_iterators.py ..............                                                                                       [ 91%]
test_smart_ptr.py .............                                                                                                      [ 93%]
test_stl.py .........s.............                                                                                                  [ 96%]
test_stl_binders.py .........                                                                                                        [ 97%]
test_tagbased_polymorphic.py .                                                                                                       [ 98%]
test_thread.py ..                                                                                                                    [ 98%]
test_union.py .                                                                                                                      [ 98%]
test_virtual_functions.py ..........                                                                                                 [100%]

========================================================= short test summary info ==========================================================
SKIPPED [1] test_eigen_matrix.py:718: could not import 'scipy': No module named 'scipy'
SKIPPED [1] test_eigen_matrix.py:728: could not import 'scipy': No module named 'scipy'
SKIPPED [1] test_stl.py:143: no <experimental/optional>
===================================================== 695 passed, 3 skipped in 10.04s ======================================================
```

Reply:
I also triggered CI testing under #4307, after git rebase master.